tests/pos/i13344.scala

```diff
@@ -0,0 +1,18 @@
+  type Nat = [A] => (A => A) => A => A
+
+  def Zero: Nat =
+    [A] => (f: A => A) => (a: A) => a
+
+  def Succ[A](n: Nat): Nat =
+    [A] => (f: A => A) => (a: A) => f(n(f)(a))
+
+  def one: Nat =
+    Succ(Zero)
+
+  def mul(m: Nat)(n: Nat): Nat =
+    [A] => (f: A => A) => (a: A) => m(n(f))(a)
+
+  def exp(m: Nat)(n: Nat): Nat =
+    // the compiler throws a StackOverflowError without the type annotations!
+    [A] => (f: A => A) => (a: A) => (n[Nat](mul(m))(one): Nat)(f)(a)
```

Comment:
I believe the reproducer has been posted in an unfortunate format -- not sure if you saw the comment
I only get the stack overflow on 3.0.2-RC1 when I change the last line to

`[A] => (f: A => A) => (a: A) => (n(mul(m))(one))(f)(a)`

(that is, when I remove the type annotations at the absolute rhs

Reply:
Ooops, I get an AssertionError, not a stack overflow error, sorry.

Reply:
Indeed, this is still broken on master then.